pkg/datapath/linux/ipsec/ipsec_linux.go

```diff
@@ -752,18 +752,60 @@ func ipsecDeleteXfrmState(nodeID uint16) error {
 		return err
 	}
 
-	errs := resiliency.NewErrorSet(fmt.Sprintf("failed to delete node (%d) xfrm states", nodeID), len(xfrmStateList))
+	xfrmStatesToDelete := []netlink.XfrmState{}
 	for _, s := range xfrmStateList {
 		if matchesOnNodeID(s.Mark) && ipsec.GetNodeIDFromXfrmMark(s.Mark) == nodeID {
-			if err := netlink.XfrmStateDel(&s); err != nil {
-				errs.Add(fmt.Errorf("failed to delete xfrm state (%s): %w", s.String(), err))
-			}
+			xfrmStatesToDelete = append(xfrmStatesToDelete, s)
+		}
+	}
+
+	errs := resiliency.NewErrorSet(fmt.Sprintf("failed to delete node (%d) xfrm states", nodeID), len(xfrmStateList))
+	for _, s := range xfrmStatesToDelete {
+		if err := safelyDeleteXfrmState(&s, xfrmStateList); err != nil {
+			errs.Add(fmt.Errorf("failed to delete xfrm state (%s): %w", s.String(), err))
 		}
 	}
 
 	return errs.Error()
 }
 
+func safelyDeleteXfrmState(state *netlink.XfrmState, stateList []netlink.XfrmState) (err error) {
+
+	if getDirFromXfrmMark(state.Mark) == dirIngress && ipsec.GetNodeIDFromXfrmMark(state.Mark) != 0 {
+		oldXFRMInMark := &netlink.XfrmMark{
+			Value: linux_defaults.RouteMarkDecrypt,
+			Mask:  linux_defaults.IPsecMarkBitMask,
+		}
+
+		errs := resiliency.NewErrorSet("failed to delete old xfrm states", len(stateList))
+
+		scopedLog := log.WithFields(logrus.Fields{
+			logfields.SPI:              state.Spi,
+			logfields.SourceIP:         state.Src,
+			logfields.DestinationIP:    state.Dst,
+			logfields.TrafficDirection: getDirFromXfrmMark(state.Mark),
+			logfields.NodeID:           getNodeIDAsHexFromXfrmMark(state.Mark),
+		})
+
+		for _, s := range stateList {
+			if s.Spi == state.Spi && xfrmIPEqual(s.Dst, state.Dst) && xfrmMarkEqual(s.Mark, oldXFRMInMark) {
+				err, deferFn := xfrmTemporarilyRemoveState(scopedLog, s, string(dirIngress))
```

Comment:
It's not clear to me why we have to remove the "old" general state ourselves first before adding it back. Why can't we just ensure that it is still there after we delete the state we actually wanted to delete?

Reply:
Because when old and new states coexist, kernel deletes old one even if cilium sends netlink message to remove new one.

Reply:
My question is why we start removing it intentionally then. I'm proposing to skip the extra delete operation if the kernel will do it anyway, e.g.:
1. Delete state with specific mark (kernel will also delete the state with the generic mark)
2. Recreate state with generic mark, if or after kernel deleted it unintentionally.

Reply:
I sort of understand what you are talking. The problem is:

> kernel will also delete the state with the generic mark

Kernel doesn't **ALSO** delete the generic state, kernel **ONLY** delete the generic state instead of the specific one.

Reply:
Ah, I didn't catch that from the commit message. I wrongly assumed it removes both. Thank you for clarifying! I think it could be worth adding that fact as a comment to the `safelyDeleteXfrmState()` function, or at least inline above the temporary removal of the generic state.